File: src/main/java/br/unifor/costify/application/factory/IngredientFactory.java
package br.unifor.costify.application.factory;

import br.unifor.costify.domain.contracts.IdGenerator;
import br.unifor.costify.domain.entity.Ingredient;
import br.unifor.costify.domain.valueobject.Id;
import br.unifor.costify.domain.valueobject.Unit;
import org.springframework.stereotype.Component;

@Component
public class IngredientFactory {
  private final IdGenerator idGenerator;

  public IngredientFactory(IdGenerator idGenerator) {
    this.idGenerator = idGenerator;
  }

  public Ingredient create(
      String name, double packageQuantity, double packagePrice, Unit packageUnit) {
    return createIngredient(idGenerator, name, packageQuantity, packagePrice, packageUnit);
  }

  public Ingredient create(
      Id id, String name, double packageQuantity, double packagePrice, Unit packageUnit) {
    return createIngredient(id, name, packageQuantity, packagePrice, packageUnit);
  }

  private Ingredient createIngredient(
      IdGenerator idGenerator,
      String name,
      double packageQuantity,
      double packagePrice,
      Unit packageUnit) {
    return new Ingredient(idGenerator, name, packageQuantity, packagePrice, packageUnit);
  }

  private Ingredient createIngredient(
      Id id, String name, double packageQuantity, double packagePrice, Unit packageUnit) {
    return new Ingredient(id, name, packageQuantity, packagePrice, packageUnit);
  }
}

File: internal/application/factory/ingredient_factory.go
package factory

import (
	"github.com/vini/costify-go/internal/application/errors"
	"github.com/vini/costify-go/internal/domain/contract"
	"github.com/vini/costify-go/internal/domain/entity"
	"github.com/vini/costify-go/internal/domain/valueobject"
)

// IngredientFactory provides factory methods for creating ingredients
type IngredientFactory struct {
	idGenerator contract.IdGenerator
}

// NewIngredientFactory creates a new IngredientFactory
func NewIngredientFactory(idGenerator contract.IdGenerator) *IngredientFactory {
	return &IngredientFactory{
		idGenerator: idGenerator,
	}
}

// Create creates an ingredient with generated ID
func (f *IngredientFactory) Create(
	name string,
	packageQuantity float64,
	packagePrice float64,
	packageUnit string,
) (*entity.Ingredient, error) {
	money, err := valueobject.Money{}.Of(packagePrice)
	if err != nil {
		return nil, err
	}

	// Defensive programming: validate unit even though HTTP layer may have validated it
	// This ensures the factory remains independent and robust
	unit, exists := valueobject.FromString(packageUnit)
	if !exists {
		return nil, errors.NewInvalidUnitError("Invalid unit: " + packageUnit)
	}

	return entity.NewIngredientWithGenerator(
		f.idGenerator,
		name,
		packageQuantity,
		money,
		unit,
	)
}

// CreateWithId creates an ingredient with explicit ID
func (f *IngredientFactory) CreateWithId(
	id valueobject.Id,
	name string,
	packageQuantity float64,
	packagePrice float64,
	packageUnit valueobject.Unit,
) (*entity.Ingredient, error) {
	money, err := valueobject.Money{}.Of(packagePrice)
	if err != nil {
		return nil, err
	}

	return entity.NewIngredient(
		id,
		name,
		packageQuantity,
		money,
		packageUnit,
	)
}
